apps/prairielearn/src/question-servers/freeform.ts

```diff
@@ -865,10 +965,17 @@ async function processQuestionHtml(phase, codeCaller, data, context) {
   } = await processFunction(...args);
 
   if (phase === 'grade' || phase === 'test') {
+    // This aligns with `PartialScore` from question_utils.py
+    // feedback: NotRequired[str | dict[str, str] | Any]
+    const partial_scores: Record<
+      string,
+      { score?: number; weight?: number; feedback?: string | object }
+    > = resultData.partial_scores ?? {};
```

Comment:
This feels like we're completely lying to ourselves. If anything, I'd rather type it as `unknown` (or at the very least, `Record<string, any>`) and force ourselves to do necessary type checks.

Reply:
It would be nice to be able to fully type resultData in such a way that the types are shared between javascript and python. For example, maybe we have a single source of truth for both.

Reply:
It would be nice to have a full type validation helper so that we can ingest these safely. Maybe Zod in a follow up PR?

Reply:
I took a pass at improving things here: 2271cfe

IMO this is worth me splitting into a separate PR.

Reply:
I moved this to #11834.